File: etrace-common/src/main/java/io/etrace/common/util/RpcIdUtil.java
/*
 * Copyright 2020 etrace.io
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.etrace.common.util;

import com.google.common.collect.Lists;
import io.etrace.common.RequestIdInfo;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Stack;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class RpcIdUtil {

    final static Pattern PAT = Pattern.compile("([.^]+)");

    public static Comparator<String> getComparatorAsString() {
        return (o1, o2) -> {
            if (o1.equals(o2)) {
                return 0;
            }
            String[] parts1 = parseToParts2(o1);
            String[] parts2 = parseToParts2(o2);

            // first, compare same part
            int sameLength = Math.min(parts1.length, parts2.length);
            for (int i = 0; i < sameLength; i++) {
                if (!parts1[i].equals(parts2[i])) {
                    try {
                        return Integer.parseInt(parts1[i]) - Integer.parseInt(parts2[i]);
                    } catch (NumberFormatException e) {
                        // ignore
                    }
                }
            }

            // then, compare the remaining length: (parts1.length - sameLength) - (parts2.length - sameLength);
            return parts1.length - parts2.length;
        };
    }

    public static Comparator<RequestIdInfo> getComparator() {
        return (o1, o2) -> {
            if (o1.getRpcId().equals(o2.getRpcId())) {
                return 0;
            }
            String[] parts1 = parseToParts2(o1.getRpcId());
            String[] parts2 = parseToParts2(o2.getRpcId());

            // first, compare same part
            int sameLength = Math.min(parts1.length, parts2.length);
            for (int i = 0; i < sameLength; i++) {
                if (!parts1[i].equals(parts2[i])) {
                    try {
                        return Long.compare(Long.parseLong(parts1[i]), Long.parseLong(parts2[i]));
                    } catch (NumberFormatException e) {
                        // ignore
                    }
                }
            }

            // then, compare the remaining length: (parts1.length - sameLength) - (parts2.length - sameLength);
            return parts1.length - parts2.length;
        };
    }

    public static String[] parseToParts2(String rpcId) {
        String[] parts = rpcId.split("[.^]+");

        if (parts.length == 1) {
            return parts;
        } else {
            List<String> result = Lists.newArrayList();
            List<String> regexParts = matchAll(PAT, rpcId);

            int index = 0;
            boolean meetRemoveCall = false;
            Stack<String> stack = new Stack<>();

            while (index < parts.length) {
                if (index < regexParts.size() && "^".equals(regexParts.get(index))) {
                    stack.push(parts[index]);
                    meetRemoveCall = true;
                } else {
                    if (meetRemoveCall) {
                        stack.push(parts[index]);

                        result.addAll(popAndClear(stack));
                        meetRemoveCall = false;
                    } else {
                        result.add(parts[index]);
                    }
                }
                index++;
            }
            return result.toArray(new String[0]);
        }
    }

    public static List<String> matchAll(Pattern pat, String input) {
        List<String> result = Lists.newArrayList();
        Matcher matcher = pat.matcher(input);

        while (matcher.find()) {
            result.add(matcher.group());
        }
        return result;
    }

    private static <T> List<T> popAndClear(Stack<T> stack) {
        List<T> result = Lists.newArrayListWithCapacity(stack.size());
        while (!stack.empty()) {
            result.add(stack.pop());
        }
        return result;
    }

    // not used
    @Deprecated
    public static String[] parseToParts(String rpcId) {
        if (rpcId.contains("^")) {
            int index = 0;
            List<String> list = Lists.newArrayList();
            Stack<String> angleStack = new Stack<>();

            while (index < rpcId.length()) {
                int nextPeriod = rpcId.indexOf(".", index);
                int nextAngle = rpcId.indexOf("^", index);

                int nextPartStart;

                if (nextPeriod < 0 && nextAngle > 0) {
                    angleStack.add(rpcId.substring(index, nextAngle));
                    nextPartStart = nextAngle;
                } else if (nextPeriod > 0 && nextAngle < 0) {
                    list.add(rpcId.substring(index, nextPeriod));
                    nextPartStart = nextPeriod;
                } else if (nextPeriod < nextAngle) {
                    if (!angleStack.empty()) {
                        list.addAll(new ArrayList<>(angleStack));
                        angleStack.clear();
                    }
                    list.add(rpcId.substring(index, nextPeriod));
                    nextPartStart = nextPeriod;
                } else {
                    // nextPeriod > nextAngle
                    // handle "^"
                    angleStack.push(rpcId.substring(index, nextAngle));
                    nextPartStart = nextAngle;
                }

                // handle "^^" case
                while (nextPartStart + 1 < rpcId.length()
                    && rpcId.charAt(nextPartStart + 1) != '.'
                    && rpcId.charAt(nextPartStart + 1) != '^') {
                    nextPartStart++;
                }

                int nextNextPeriod = rpcId.indexOf(".", nextPartStart + 1);
                int nextNextAngle = rpcId.indexOf("^", nextPartStart + 1);

                if (nextNextAngle < 0 && nextNextPeriod < 0) {
                    list.add(rpcId.substring(index));
                    index = rpcId.length();
                } else if (nextNextPeriod < 0) {
                    index = nextNextAngle;
                } else if (nextNextAngle < 0) {
                    index = nextNextPeriod;
                } else {
                    index = Math.min(nextNextAngle, nextNextPeriod);
                }
            }

            if (!angleStack.empty()) {
                list.addAll(new ArrayList<>(angleStack));
                angleStack.clear();
            }
            return list.toArray(new String[0]);
        } else {
            return rpcId.split("[.^]");
        }
    }
}

File: etrace-ui/src/utils/RpcIdUtil.ts
import {Stack} from "immutable";

export default class RpcIdUtil {
    private static pat: RegExp = /([.^]+)/g;

    /**
     * 可能的CASE 有： 1.2.3^4  => 1.2.4, 1^2 => 2
     * 详见 Callstack.test.tsx
     * @param rpcId
     */
    public static parseFatherRpcId(rpcId: string): string {
        let dot = rpcId.lastIndexOf(".");
        let upper = rpcId.lastIndexOf("^");

        let index = dot > upper ? dot : upper;

        if (upper > 0 && upper > dot) {
            // 这是最后为^的 rpcId
            let lastPart = rpcId.substring(upper + 1);
            if (dot > 0) {
                let firstPart = rpcId.substring(0, dot);
                return firstPart + "." + lastPart;
            } else {
                return lastPart;
            }
        } else {
            return rpcId.substr(0, index);
        }
    }

    /*
     depth first sort
     return negative if the first item is smaller; positive if it it's larger, or zero if they're equal.
     */
    public static dfs(first: string, second: string): number {
        if (first == second) {
            return 0;
        } else {
            let parts1: Array<string> = this.parseToParts2(first);
            let parts2: Array<string> = this.parseToParts2(second);
            // first, compare same part
            let sameLength: number = Math.min(parts1.length, parts2.length);

            for (let i = 0; i < sameLength; i++) {
                if (parts1[i] != (parts2[i])) {
                    try {
                        let i1 = parseInt(parts1[i]);
                        let i2 = parseInt(parts2[i]);

                        // 让 NONE.1 == 1.1
                        if (!isNaN(i1) && !isNaN(i2)) {
                            return i1 - i2;
                        }
                    } catch (e) {
                        console.warn(first, second, e);
                        // ignore
                    }
                }
            }
            // then, compare the remaining length: (parts1.length - sameLength) - (parts2.length - sameLength);
            return parts1.length - parts2.length;
        }
    }

    /*
    breadth first sort
    return negative if the first item is smaller; positive if it it's larger, or zero if they're equal.
    */
    public static bfs(first: string, second: string): number {
        if (first == second) {
            return 0;
        } else {
            let parts1: Array<string> = this.parseToParts2(first);
            let parts2: Array<string> = this.parseToParts2(second);

            if (parts1.length != parts2.length ) {
                return parts1.length - parts2.length;
            } else {
                // first, compare same part
                let sameLength: number = parts1.length;

                for (let i = 0; i < sameLength; i++) {
                    if (parts1[i] != (parts2[i])) {
                        try {
                            let i1 = parseInt(parts1[i]);
                            let i2 = parseInt(parts2[i]);

                            if (!isNaN(i1) && !isNaN(i2)) {
                                return i1 - i2;
                            }
                        } catch (e) {
                            console.warn(first, second, e);
                            // ignore
                        }
                    }
                }
                return 0;
            }
        }
    }

    public static parseToParts2(rpcId: string): Array<string> {
        let parts: Array<string> = rpcId.split(/[.^]+/);
        if (parts.length == 1) {
            return parts;
        } else {
            let result: Array<string> = [];
            let regexParts: Array<string> = rpcId.match(RpcIdUtil.pat);

            let index: number = 0;
            let meetRemoveCall: boolean = false;
            // 默认是 immutable的，需要加上asMutable()
            let stack: Stack<{}> = Stack.of().asMutable();

            while (index < parts.length) {
                if (index < regexParts.length && regexParts[index] == "^") {
                    stack.push(parts[index]);
                    meetRemoveCall = true;
                } else {
                    if (meetRemoveCall) {
                        stack.push(parts[index]);
                        result.push(...this.popAndClear(stack));
                        meetRemoveCall = false;
                    } else {
                        result.push(parts[index]);
                    }
                }
                index++;
            }
            return result;
        }
    }

    private static popAndClear(stack: Stack<{}>): Array<string> {
        let popped = [];

        while (stack.size != 0) {
            // 使用的这个 immutable package中的stack api比较奇怪
            popped.push(stack.peek());
            stack = stack.shift();
        }
        return popped;
    }
}
